Use GitLab for syncing upstream->downstream - gitlab_git_changes

phracek opened the pull request:
This pull request allows using gitlab for
syncing upstream->downstream.

The rest of the function remains. Like
git clonedownstream uses dist-git.

This pull request covers 'git pullupstream', 'git show', and 'git push'.
The GitLab functionality is valid only for RHEL and RHCWT especially. CWT itself does not make sense for this.
The question is if 'gitlab.py' should be in CWT or should be moved to RHCWT, like 'bugzilla.py' or 'jira.py'

The calling is like this:

```bash
$ cwt -v 5 --config=rhel8.yaml:rhel8.6 --base <base_image> --do-image <image_name> git pullupstream --gitlab
$ cwt -v 5 --config=rhel8.yaml:rhel8.6 --base <base_image> --do-image <image_name> git show --gitlab
$ cwt -v 5 --config=rhel8.yaml:rhel8.6 --base <base_image> --do-image <image_name> git push --gitlab
```

Signed-off-by: Petr "Stone" Hracek <phracek@redhat.com>

phracek commented:
/test

phracek commented:
/test

phracek commented:
/test

phracek commented:
/test

pkubatrh commented:
Before I get into the details I have a sugggestion - how about making the gitlab module an extension of the distgit module and only modifying what is necessary to change? I expect a lot of the functionality would stay the same (just with different server and APIs) or could be changed to be more configurable.

phracek commented:
> Before I get into the details I have a suggestion - how about making the gitlab module an extension of the distgit module and only modifying what is necessary to change? I expect a lot of the functionality would stay the same (just with different server and APIs) or could be changed to be more configurable.

Thanks for this comment.

I think more functions like `_get_from`, `_set_from`, `update_dockerfile`, etc. can go into the alone class.
Function like '_pull_upstream' can be moved into a separate class, even with static function as well.
This can be done as a separate pull request, so it is not complicated for review.

We can make GitLab module on the top of distgit module, but I would prefer to use dist-git module only for cloning downstream (dist-git), building, and GitLab module for operation with upstream -> downstream.

phracek commented:
@pkubatrh Feel free to review it a bit.

phracek commented:
/test

phracek commented:
The pull request can be reviewed. I would be glad for you feedback.
@pkubatrh @frenzymadness 

/test

frenzymadness reviewed (commented):
I did a first quick scan. I won't have time for that sooner than after devconf.

phracek commented:
The description of this pull request has been updated with CLI examples.

/test

phracek commented:
gitlab.py was moved to internal repository. There are some integrations like Slack, that are private.

/test

phracek commented:
gitlab.py was moved to internal repository. There are some integrations like Slack, that are private.

/test

frenzymadness commented on container_workflow_tool/git_operations.py:
```diff
@@ -113,20 +113,24 @@ def _clone_upstream(self, url, ups_path, commands=None):
         select = "origin/" + branch + ".." + branch
+        if branch_name != "":
+            select = "origin/" + branch_name + ".." + branch
         if len(list(repo.iter_commits(select))) == 0:
```
You can shorten this condition into:

```suggestion
        return bool(list(repo.iter_commits(select)))
```

It returns False for empty list and True otherwise.

But docstring says that this function should return a list of commits or emty array, which is no longer true.

frenzymadness commented on container_workflow_tool/git_operations.py:
```diff
@@ -135,6 +139,8 @@ def show_git_changes(self, tmp, components=None, diff=False):
             tmp (str): Path to the directory that is used to store git repositories
             components (list of str, optional): List of components to show changes for
             diff (boolean, optional): Controls whether the method calls git-show or git-diff
+            branch_name (str, optional): In case of gitlab, forked_branch and branch_name has to be defined.
```
`forked_branch` is mentioned multiple times in this file but there is no such variable or function argument.

phracek replied:
Fixeds by https://github.com/sclorg/container-workflow-tool/pull/71/commits/f62abfafd01a520ef7d759971344db80b36a0ce6

frenzymadness commented on container_workflow_tool/main.py:
```diff
@@ -117,6 +117,8 @@ def _setup_args(self, args):
             self.set_commit_msg(args.commit_msg)
         if getattr(args, 'rebuild_reason', None) is not None and args.rebuild_reason:
             self.rebuild_reason = args.rebuild_reason
+        if getattr(args, 'gitlab', None) is not None and args.gitlab:
```
First of all, these two lines seem to be here twice, see lines 130 and 131.

And IMHO, if an argument uses `store_true` or `store_false` actions, it has to be True or False, nothing else, which means that the check for the attribute presence and the default None here are not needed. The argument will always be there.

phracek replied:
Some options are valid only for a specific subcommand, like gitlab. But I will check what can be enhanced. You are right.

frenzymadness commented on container_workflow_tool/main.py:
```diff
@@ -521,24 +522,19 @@ def pull_upstream(self):
-            self.distgit._clone_upstream(i["git_url"],
-                                         ups_name,
-                                         commands=i["commands"])
+            self.git_ops.clone_upstream(i["git_url"], ups_name, commands=i["commands"])
```
Would be nice to have a better name than `i` here.

phracek replied:
Fixed by https://github.com/sclorg/container-workflow-tool/pull/71/commits/d1623aad6649f01099cadc67adbb67c6ed8df53a

frenzymadness commented on container_workflow_tool/main.py:
```diff
@@ -583,13 +570,23 @@ def dist_git_changes(self, rebase: bool = False):
                 "cwt git push && cwt build"
                 "[base/core/s2i] --repo-url link-to-repo-file")
 
+    def dist_git_changes(self, rebase: bool = False):
```
What about
```suggestion
    def dist_git_merge_changes(self, rebase: bool = False):
```

phracek replied:
Fixed by commit https://github.com/sclorg/container-workflow-tool/pull/71/commits/80e57ad69a6587abc547a99c6632d6d34665f2aa.

The version was also bumped because there are none-compatible changes.

frenzymadness commented on container_workflow_tool/utility.py:
```diff
@@ -73,6 +74,22 @@ def _split_config_path(config: str):
+@contextlib.contextmanager
+def cwd(path):
+    """
+    Checks out a git repository into a temporary directory.
```
This line no longer applies.

frenzymadness reviewed (changes requested):
Next round of review. I'm not a user of this tool so I might miss the big picture but I spot some potential improvements.

phracek commented on container_workflow_tool/utility.py:
```diff
@@ -73,6 +74,22 @@ def _split_config_path(config: str):
+    Checks out a git repository into a temporary directory.
+    Changes CWD to the temporary directory.
+    Yields the temporary directory.
+    On exit, the temporary directory is removed and CWD is restored.
```
fixed by https://github.com/sclorg/container-workflow-tool/pull/71/commits/d1623aad6649f01099cadc67adbb67c6ed8df53a

phracek commented on container_workflow_tool/distgit.py:
```diff
@@ -63,7 +63,7 @@ def dist_git_changes(self, images, rebase=False):
         """
         try:
-            for image in (images):
+            for image in images:
```
Fixed by https://github.com/sclorg/container-workflow-tool/pull/71/commits/d1623aad6649f01099cadc67adbb67c6ed8df53a

phracek commented:
/test

phracek commented:
@pkubatrh @frenzymadness Hi folks, I have addressed all issues, If you have nothing against merging, I would like to merge it, so I can work on RHCWT.

frenzymadness reviewed (approved):
I think it's good to go.
